exammode/models.py

```diff
@@ -0,0 +1,107 @@
+from django.db import models
+from django.urls import reverse
+from django.utils import timezone
+
+from course.models import CourseInstance, CourseModule
+from exercise.exercise_models import LearningObject
+from userprofile.models import UserProfile
+
+# Create your models here.
+
+
+class ActiveExamSessionManager(models.Manager):
+
+    def get_queryset(self):
+        initial_queryset = super().get_queryset()
+        queryset = [q for q in initial_queryset if (
+            q.can_start <= timezone.now())]
+
+        return queryset
+
+    def is_active(self):
+        initial_queryset = super().get_queryset()
+        queryset = [q for q in initial_queryset if (q.can_start <= timezone.now() and (
+            timezone.now() <= q.can_start + timezone.timedelta(hours=q.duration)))]
+
+        return queryset
+
+
+class ExamSession(models.Model):
+    """
+    Represents one instance of a course exam
+    """
+
+    course_instance = models.ForeignKey(
+        CourseInstance, on_delete=models.CASCADE)
+    exam_module = models.ForeignKey(
+        CourseModule, on_delete=models.CASCADE, null=True, blank=True)
+    can_start = models.DateTimeField(editable=True, default=timezone.now)
+    duration = models.IntegerField()
+    room = models.CharField(max_length=255, null=True, blank=True)
+
+    objects = models.Manager()
+    active_exams = ActiveExamSessionManager()
+
+    def __str__(self):
+        return " ".join([str(self.course_instance), str(self.can_start)])
+
+    def start_exam(self, user):
+
+        # Checking first if exam content is available. If not, database entries would be pointless
+        learning_objects = LearningObject.objects.filter(
+            course_module__exact=self.exam_module
+        )
+
+        if learning_objects:
+            redirect_url = learning_objects[0].get_display_url()
+
+        else:
+            return reverse("exam_module_not_defined")
+
+        attempt = ExamAttempt(
+            exam_taken=self,
+            student=user.userprofile,
+            exam_started=timezone.now()
+        )
+        attempt.save()
+
+        user.userprofile.active_exam = attempt
+        user.userprofile.save()
+
+        return redirect_url
+
+    def end_exam(self, user):
+        attempt = ExamAttempt.objects.filter(
+            exam_taken=self, student=user.userprofile)[:1].get()
```

Comment:
The `first` method is better that `[:1].get()`: https://docs.djangoproject.com/en/2.2/ref/models/querysets/#first

Reply:
There is no need to filter the exam attemps, since there can be only one active exam attempt per user. Attempt can be search simply with `user.userprofile.active_exam`.